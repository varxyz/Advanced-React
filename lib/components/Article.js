import React from 'react';
import PropTypes from 'prop-types';
const dateDisplay = (dateString) => new Date(dateString).toDateString();

const Article = ({ article }, context) => {

  /**
   * context will become available as the second argument in
   * a functional component, so now we can read the store from
   * the context instead of the props
   */

  const author = context.store.lookupAuthor(article.authorId);

  return (
    <div>
      <div>{article.title}</div>
      <div>{dateDisplay(article.date)}</div>
      <div>
        <a href={author.website}>
          {author.firstName} {author.lastName}
        </a>
      </div>
      <div>{article.body}</div>
    </div>
  );
};

Article.propTypes = {
  article: PropTypes.shape({
    date: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired
  })
};

/**
 * In order to allow this component to access the context, we
 * need to define the `contextTypes` first
 */

Article.contextTypes = {
  store: PropTypes.object
};

export default Article;
