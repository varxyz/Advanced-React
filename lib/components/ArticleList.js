import React from 'react';
import Article from './Article';

/**
 * articleList doesnt need to rerender everytime,but function components
 * in react render regardless, they are not optimized perf wise. Unless
 * we dont use a HOC library like react-redux that wraps the function
 * component and optimizes it, we have 2 solutions for this --
 * One solution will be to call this function in the parent `App`
 * as {ArticleList()} the usual way instead of component style
 * like <ArticleList/>
 * The second option is to use a PureComponent instead of a function
 */

class ArticleList extends React.PureComponent {
  render() {
    return (
      <div>
        {Object.values(this.props.articles).map((article) => (
          <Article
            key={article.id}
            article={article}
          />
        ))}
      </div>
    );
  }
}

export default ArticleList;
