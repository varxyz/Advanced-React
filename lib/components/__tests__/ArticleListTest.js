import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Article from '../Article';

configure({ adapter: new Adapter() });

import ArticleList from '../ArticleList';

describe('ArticleList', () => {
  const testProps = {
    articles: {
      a: {id: 2},
      b: {id: 1}
    },
    /**
     * WE no longer pass the store down to the article comp
     * through props, but access it directly through context
     */

    // store: {
    //   lookupAuthor: jest.fn(() => ({}))
    // }
  };

  Article.propTypes = {};

  it('renders correctly', () => {
    /**
     * with shallow we just render the articleList without the
     * childs (Articles)
     */
    const wrapper = shallow(
      <ArticleList
        {...testProps}
      />
    );

    expect(wrapper.find('Article').length).toBe(2);
    expect(wrapper).toMatchSnapshot();

    //   const el = renderer
    //     .create(
    //       <ArticleList
    //         {...testProps}
    //       />
    //     )
    //     .toJSON();
    //   expect(el.children.length).toBe(2);
    //   expect(el).toMatchSnapshot();
    // });
  });
});

/**
 * after adding the context object to our app, the tests fail,
 * as `ArticleList` no longer passes the store down to the
 * Article component, so the `lookupAuthor` method is now undefined.
 * We will make use of shallow rendering, to unit test just the article
 * 
 */
