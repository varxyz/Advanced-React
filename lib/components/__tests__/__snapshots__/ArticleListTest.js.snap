// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleList renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UsersList
    articles={
      Object {
        "a": Object {
          "id": 2,
        },
        "b": Object {
          "id": 1,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ArticleContainer
          article={
            Object {
              "id": 2,
            }
          }
        />,
        <ArticleContainer
          article={
            Object {
              "id": 1,
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "article": Object {
            "id": 2,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "article": Object {
            "id": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ArticleContainer
            article={
              Object {
                "id": 2,
              }
            }
          />,
          <ArticleContainer
            article={
              Object {
                "id": 1,
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "article": Object {
              "id": 2,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "article": Object {
              "id": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
